docs/technical/git-submodules.md

```diff
@@ -0,0 +1,167 @@
+submodules.
+
+This document is a work in progress; submodules are a big feature, and relevant
+details will be filled in incrementally.
+
+## Objective
+
+This proposal aims to replicate the workflows users are used to with Git
+submodules, e.g.:
+
+- Cloning submodules
+- Making new submodule commits and updating the superproject
+- Fetching and pushing updates to the submodule's remote
+- Viewing submodule history
+
+When it is convenient, this proposal will also aim to make submodules easier to
+use than Git's implementation.
+
+### Non-goals
+
+- Non-Git 'submodules' (e.g. native jj submodules, other VCSes)
+- Non-Git backends (e.g. Google internal backend)
+- Changing how Git submodules are implemented in Git
+
+## Background
+
+We mainly want to support Git submodules for feature parity, since Git
+submodules are a standard feature in Git and are popular enough that we have
+received user requests for them. Secondarily (and distantly so), Git submodules
+are notoriously difficult to use, so there is an opportunity to improve the UX
+over Git's implementation.
+
+### Git Submodules 101
+
+[Git submodules](https://git-scm.com/docs/gitsubmodules) are a feature of Git
+that allow a repository (submodule) to be embedded inside another repository
+(the superproject). Notably, a submodule is a full repository, complete with its
+own index, object store and ref store. It can be interacted with like any other
+repository, regardless of the superproject.
+
+In a superproject commit, submodule information is captured in two places:
+
+- A `gitlink` entry in the commit's tree, where the value of the `gitlink` entry
+  is the submodule commit id. This tells Git what to populate in the working
+  tree.
+
+- A top level `.gitmodules` file. This file is in Git's config syntax and
+  entries take the form `submodule.<submodule-name>.*`. These include many
+  settings about the submodules, but most importantly:
+
+  - `submodule<submodule-name>.path` contains the path from the root of the tree
+    to the `gitlink` being described.
+
+  - `submodule<submodule-name>.url` contains the url to clone the submodule
+    from.
+
+In the working tree, Git notices the presence of a submodule by the `.git` entry
+(signifying the root of a Git repository working tree). This is either the
+submodule's actual Git directory (an "old-form" submodule), or a `.git` file
+pointing to `<superproject-git-directory>/modules/<submodule-name>`. The latter
+is sometimes called the "absorbed form", and is Git's preferred mode of
+operation.
+
+## Roadmap
+
+Git submodules should be implemented in an order that supports an increasing set
+of workflows, with the goal of getting feedback early and often. When support is
+incomplete, jj should not crash, but instead provide fallback behavior and warn
+the user where needed.
+
+The goal is to land good support for pure Jujutsu repositories, while colocated
+repositories will be supported when convenient.
```

Comment:
I realize that this statement on colocation might not capture how I actually think about the roadmap. I initially wrote this thinking that the pure jj case and colocated case wouldn't be that different, so this was more of an escape hatch in case something about colocation turned out to be more difficult than expected.

But I'm not sure if this 'incremental use case' sort of roadmap makes a lot of sense without colocated repo support. I'd expect Git submodule early adopters to use colocated repos since they can fall back to Git for the use cases jj doesn't support. On the other hand, if we only handle pure jj repositories, we'd need an almost complete implementation before users can actually do anything with submodules.

What do others think? Does this roadmap make sense even if we only support jj-only?

Reply:
I would hope that simply populating the submodules in the working copy would be enough for many users, but I don't use any repos with submodules, so I don't really know. Is it common that users have to modify submodules (not just the superproject)?

Reply:
> I realize that this statement on colocation might not capture how I actually think about the roadmap. I initially wrote this thinking that the pure jj case and colocated case wouldn't be that different, so this was more of an escape hatch in case something about colocation turned out to be more difficult than expected.

I think adding the escape hatch is a good idea, it gives us an out when we're unable to progress. But your point stands that we'd get no direct feedback, when decisions heavily depend on it. So I'd say we "officially" mention the pure use-case as the final goal, while keeping colocated repositories as best effort. This allows a clean design, while getting the incremental feedback you wanted.

But I've never used submodules.

Reply:
> so I don't really know. Is it common that users have to modify submodules (not just the superproject)?

the major use-case (inside Google) would be Android development. The superproject (https://googleplex-android.git.corp.google.com/platform/superproject) is just a config that stitches together ~1000 submodules. Most work on android touches just a single submodule, but about 10% needs changes that coordinate across submodules.

Other examples that I know of are Gerrit. There we have a submodule for JGit (a base dependency), and for plugins,  (https://gerrit.googlesource.com/gerrit/+/refs/heads/master/plugins/).

JGit development typically happens in the JGit repo, but it's useful for prototyping/debugging to make changes in the submodule too.

For plugins, code refactoring in Gerrit main repo (which provides APIs) may cause changes in the plugins (consumers of the APIs), and they need to be coordinated.

Reply:
In the Gerrit scenario, would it be valuable to have readonly submodules in jj? I'd imagine it's possible that changes in the submodule don't happen that frequently, so there might be users who would use jj if it allowed them to just build the submodule.

Reply:
having readonly submodules would allow the Gerrit team dogfood JJ. Changes to plugins or JGit are uncommon enough.

Reply:
Perfect. It sounds like this roadmap still makes sense in a jj-only world. Thanks!